File: Week2/ArrayProblem/LongestSubArrayWithSumK.java
import java.util.HashMap;

class LongestSubArrayWithSumK{
/*
* Given an array containing N integers and an integer K. Your task is to find the length of the longest Sub-Array 
* with the sum of the elements equal to the given value K.
* https://www.geeksforgeeks.org/problems/longest-sub-array-with-sum-k0809/1
* T(c) -> O(N^2), S(c) -> O(1)
 */
    static int lenOfLongSubarr_BruteForce(int A[], int N, int K) {
        int sum = 0;
        int maxLen = 0;
        
        for(int i=0;i<N;i++)
        {
            for(int j=i;j<N;j++)
            {
                sum += A[j];
                int currentSubarrLen = (j - i) + 1;
                if(sum == K && currentSubarrLen > maxLen)
                {
                    maxLen = currentSubarrLen;
                }
            }
            /* reset sum for finding next subarray */
            sum = 0;
        }
        return maxLen;
    }


    /* 
    * T(c) -> O(N), S(c) -> O(N)
     */
    public static int lenOfLongSubarr_Efficient(int A[], int N, int K) {
        /* using hashing */
        /* hashmap will store sum with index */
        HashMap<Integer, Integer> map = new HashMap<>();
        int sum = 0;
        int maxLen = 0;
        
        for(int i=0;i<N;i++)
        {
            sum += A[i];
            /* when we will find the first subarray, calculate that length */
            if(sum == K)
            {
                maxLen = Math.max(maxLen, i+1);
            }
            
            /* calculate sum of remaining part, x-k */
            int prefixSum = sum - K;
            if(map.containsKey(prefixSum))
            {
                int len = i - map.get(prefixSum);
                maxLen = Math.max(maxLen, len);
            }
            
            /* if sum is not present in map, update that */
            if(!map.containsKey(sum))
            {
                map.put(sum,i);
            }
        }
        return maxLen;
    }


    public static void main(String[] args) {
        int arr[] = {10, 5, 2, 7, 1, 9};
        int k = 15;
        System.out.println(lenOfLongSubarr_BruteForce(arr, arr.length, k));
        int arr1[] = {1, 4, 3, 3, 5, 5};
        k = 16;
        System.out.println(lenOfLongSubarr_Efficient(arr1, arr1.length, k));
    }
}

File: Week2/Python/ArrayProblem/LongestSubArrayWithSumK.py
"""
    Given an array arr containing n integers and an integer k. Your task is to find the length of the longest Sub-Array 
    with the sum of the elements equal to the given value k.

    https://www.geeksforgeeks.org/problems/longest-sub-array-with-sum-k0809/1
"""


"""Time limit issue"""
def lenOfLongSubarr(arr, n, k) : 
    """ T(c) -> O(n^2), S(c) -> O(1) """

    max_len = 0
    
    for i in range(n):
        total = 0
        
        for j in range(i, n):
            total += arr[j]
            
            # if subarray sum is equal to k, check if that is maxlen
            if total == k:
                max_len = max(max_len, j-i+1)

    return max_len



def lenOfLongSubarrEfficeint(arr, n, k) : 
    """ T(c) -> O(n), S(c) -> O(n) """

    # it stores commulative sum and it's index
    prefix_map = {}
    max_len = 0
    current_sum = 0
    
    for i in range(n):
        current_sum += arr[i]
        
        # if subarray sum is equal to k
        if current_sum == k:
            max_len = i+1
        
        # calculate prefix sum
        prefix_sum = current_sum - k
        
        # if prefix sum exist
        if prefix_sum in prefix_map:
            curr_len = i - prefix_map[prefix_sum]
            max_len = max(max_len, curr_len)
        
        # if current sum not exist in map, add it
        if current_sum not in prefix_map:
            prefix_map[current_sum] = i

    return max_len



n = 6
arr = [10, 5, 2, 7, 1, 9]
k = 15
print(lenOfLongSubarr(arr, n,  k))

print(lenOfLongSubarrEfficeint(arr, n, k))
